content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,95 @@
+
+== Using shared libraries
+
+=== Do not override built-in Pipeline Steps
+
+Wherever possible stay away from customized/overwritten Pipeline steps. 
+Overriding built-in Pipeline Steps is the process of using shared libraries to overwrite the standard Pipeline APIs like `sh` or `timeout`. 
+This process is dangerous because the Pipeline APIs can change at any time causing custom code to break or give different results than expected. 
+When custom code breaks because of Pipeline API changes, troubleshooting is difficult because even if the custom code has not changed, it may not work the same after an API update.
+So even if custom code has not changed that does not mean that after an API update it will keep working the same. 
+Lastly, because of the ubiquitous use of these steps throughout Pipelines, if something is coded incorrectly/inefficiently the results can be catastrophic to Jenkins.
+
+=== Avoiding large global variable declaration files
+
+Having large variable declaration files can require large amounts of memory for little to no benefit, because the file is loaded for every Pipeline whether the variables are needed or not. Creating small variable files that contain only variables relevant to the current execution is recommended.
+It would be better to load variables which are relevant to the current execution.
+
+=== Avoiding very large shared libraries
+
+Using large shared libraries in Pipelines requires checking out a very large file before the Pipeline can start and loading the same shared library per job that is currently executing, which can lead to increased memory overhead and slower execution time.
+
+== Answering additional FAQs
+
+=== Dealing with Concurrency in Pipelines
+
+Try not to share workspaces across multiple Pipeline executions or multiple distinct Pipelines. 
+This practice can lead to either unexpected file modification within each Pipeline or workspace renaming.
+
+Ideally, shared volumes/disks are mounted in a separate place and the files are copied from that place to the current workspace.
+Then when the build is done the files can be copied back if there were updates done.
+
+Build in distinct containers which create needed resources from scratch(cloud-type agents work great for this). 
+Building these containers will ensure that the build process begins at the start every time and is easily repeatable.
+If building containers will not work, disable concurrency on the Pipeline or use the Lockable Resources Plugin to lock the workspace when it is running so that no other builds can use it while it is locked.
+**WARNING**: Disabling concurrency or locking the workspace when it is running can cause Pipelines to become blocked when waiting on resources if those resources are arbitrarily locked.
+
+**Also be aware that both of these slow down the time to results of builds over being able to use unique resources for each job**
```

Comment:
Also,  be aware that both of these methods have slower time to results of builds than using unique resources for each job